src/tasks/EmitBundleTask/EmitBundleBase.cs

```diff
@@ -0,0 +1,400 @@
+            GenerateBundledResourcesRegistration(BundleRegistrationFunctionName, outputUtf8Writer);
+        });
+
+        return true;
+    }
+
+    public void Cancel()
+    {
+        BuildTaskCancelled.Cancel();
+    }
+
+    #region Helpers
+
+    private static readonly Encoding Utf8NoBom = new UTF8Encoding(false);
+    private static readonly byte[] HexToUtf8Lookup = InitLookupTable();
+    private static readonly byte[] NewLineAndIndentation = new[] { (byte)0x0a, (byte)0x20, (byte)0x20 };
+
+    private static byte[] InitLookupTable()
+    {
+        // Every 6 bytes in this array represents the output for a different input byte value.
+        // For example, the input byte 0x1a (26 decimal) corresponds to bytes 156-161 (26*6=156),
+        // whose values will be ['0', 'x', '1', 'a', ',', ' '], which is the UTF-8 representation
+        // for "0x1a, ". This is just a faster alternative to calling .ToString("x2") on every
+        // byte of the input file and then pushing that string through UTF8Encoding.
+        var lookup = new byte[256 * 6];
+        for (int i = 0; i < 256; i++)
+        {
+            string byteAsHex = i.ToString("x2");
+            char highOrderChar = BitConverter.IsLittleEndian ? byteAsHex[0] : byteAsHex[1];
+            char lowOrderChar = BitConverter.IsLittleEndian ? byteAsHex[1] : byteAsHex[0];
+            lookup[i * 6 + 0] = (byte)'0';
+            lookup[i * 6 + 1] = (byte)'x';
+            lookup[i * 6 + 2] = (byte)highOrderChar;
+            lookup[i * 6 + 3] = (byte)lowOrderChar;
+            lookup[i * 6 + 4] = (byte)',';
+            lookup[i * 6 + 5] = (byte)' ';
+        }
+
+        return lookup;
+    }
+
+    public abstract bool Emit(string destinationFile, Action<Stream> inputProvider);
+
+    private static Dictionary<string, int> symbolDataLen = new();
+
+    private static void GeneratePreallocatedMonoBundleStructs(StreamWriter outputUtf8Writer, ICollection<(string registeredName, string symbol, string? symfileSymbol)> files)
+    {
+        var preallocatedSymbols = new ArrayList();
+        StringBuilder preallocatedSource = new();
+
+        foreach (var tuple in files)
+        {
+            var preallocatedSymbol = new TaskItem(tuple.symbol);
+
+            // extern symbols
+            StringBuilder preallocatedData = new();
+            preallocatedData.AppendLine($"extern const unsigned char {tuple.symbol}_data[];");
+            if (!string.IsNullOrEmpty(tuple.symfileSymbol))
+            {
+                preallocatedData.AppendLine($"extern const unsigned char {tuple.symfileSymbol}_data[];");
+            }
+            preallocatedSource.AppendLine(preallocatedData.ToString());
+
+            // Generate Preloaded MonoBundled*Resource structs
+            string preloadedStruct;
+            switch (GetFileType(tuple.registeredName)) {
+            case "MONO_BUNDLED_ASSEMBLY": {
+                preloadedStruct = Utils.GetEmbeddedResource("mono-bundled-assembly.c");
```

Comment:
We will probably read these fragments multiple times, so maybe we should at least reuse them within this function if already read?

Reply:
Good point, instead of loading the embedded resource every time, will define string variables to hold each value and just initialize `preloadedStruct` to the appropriate value.